serf/serf.go

```diff
@@ -437,33 +436,34 @@ func (s *Serf) KeyManager() *KeyManager {
 }
 
 // UserEvent is used to broadcast a custom user event with a given
-// name and payload. The events must be fairly small, and if the
-// size limit is exceeded and error will be returned. If coalesce is enabled,
-// nodes are allowed to coalesce this event. Coalescing is only available
-// starting in v0.2
+// name and payload. If the configured size limit is exceeded and error will be returned.
+// If coalesce is enabled, nodes are allowed to coalesce this event.
+// Coalescing is only available starting in v0.2
 func (s *Serf) UserEvent(name string, payload []byte, coalesce bool) error {
-	// Check the size limit
-	if len(name)+len(payload) > UserEventSizeLimit {
-		return fmt.Errorf("user event exceeds limit of %d bytes", UserEventSizeLimit)
-	}
-
 	// Create a message
 	msg := messageUserEvent{
 		LTime:   s.eventClock.Time(),
 		Name:    name,
 		Payload: payload,
 		CC:      coalesce,
 	}
-	s.eventClock.Increment()
-
-	// Process update locally
-	s.handleUserEvent(&msg)
 
 	// Start broadcasting the event
 	raw, err := encodeMessage(messageUserEventType, &msg)
 	if err != nil {
 		return err
 	}
+
+	// Check the size
+	if len(raw) > s.config.UserEventSizeLimit {
+		return fmt.Errorf("user event exceeds limit of %d bytes", s.config.UserEventSizeLimit)
+	}
```

Comment:
I don't think we should move this check - if user maliciously or due to a bug passes a 1GB payload then doing all the work to allocate and encode that before returning the error is wasted.

In practice the encoding overhead is small and fixed-size so anything more than a few bytes is almost entirely dominated by the user payload (and name) so failing earlier seems better to me.

Reply:
Failing earlier :+1: